src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrPusher.cs

```diff
@@ -0,0 +1,168 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Text;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+using LibGit2Sharp;
+using Microsoft.DotNet.Darc.Models.VirtualMonoRepo;
+using Microsoft.DotNet.DarcLib.Helpers;
+using Microsoft.Extensions.Logging;
+
+#nullable enable
+namespace Microsoft.DotNet.DarcLib.VirtualMonoRepo;
+
+public interface IVmrPusher
+{
+    Task Push(string remote, string branch, bool verifyCommits, string? gitHubApiPat, CancellationToken cancellationToken);
+}
+
+public class VmrPusher : IVmrPusher
+{
+    private readonly IVmrInfo _vmrInfo;
+    private readonly ILogger _logger;
+    private readonly ISourceManifest _sourceManifest;
+    private readonly HttpClient _httpClient;
+    private readonly VmrRemoteConfiguration _vmrRemoteConfiguration;
+    private const string GraphQLUrl = "https://api.github.com/graphql";
+
+    public VmrPusher( 
+        IVmrInfo vmrInfo, 
+        ILogger logger,
+        ISourceManifest sourceManifest,
+        IHttpClientFactory httpClientFactory,
+        VmrRemoteConfiguration config)
+    {
+        _vmrInfo = vmrInfo;
+        _logger = logger;
+        _sourceManifest = sourceManifest;
+        _vmrRemoteConfiguration = config;
+        _httpClient = httpClientFactory.CreateClient("GraphQL");
+    }
+
+    public async Task Push(string remote, string branch, bool verifyCommits, string? gitHubApiPat, CancellationToken cancellationToken)
+    {
+        if (!verifyCommits ||
+            (gitHubApiPat != null && await CheckCommitAvailability(gitHubApiPat, cancellationToken)))
+        {
+            ExecutePush(remote, branch);
+        }
+        else
+        {
+            throw new Exception("Not all pushed commits are publicly available");
+        }  
+    }
+
+    private void ExecutePush(string remoteName, string branchName)
+    {
+        using var vmrRepo = new Repository(_vmrInfo.VmrPath);
+        vmrRepo.Config.Add("user.name", Constants.DarcBotName);
+        vmrRepo.Config.Add("user.email", Constants.DarcBotEmail);
+
+        var remote = vmrRepo.Network.Remotes[remoteName];
+        if (remote == null)
+        {
+            throw new Exception($"No remote named {remoteName} found in VMR repo.");
+        }
+
+        var branch = vmrRepo.Branches[branchName];
+        if (branch == null)
+        {
+            _logger.LogDebug($"No branch {branchName} found in VMR repo. Creating a new branch {branchName}");
```

Comment:
If the branch doesn't exist, we should error out. We only want to push an existing thing and if it doesn't exist, we'd expect someone to create it before calling push.